interface/ceed.c

```diff
@@ -26,7 +26,9 @@ static struct {
   int (*init)(const char *resource, Ceed f);
   unsigned int priority;
 } backends[32];
+CeedPragmaThreadPrivate(backends)
 static size_t num_backends;
+CeedPragmaThreadPrivate(num_backends)
```

Comment:
Hmm, does this mean every thread needs to call `CeedInit`? That feels inflexible compared to being able to create multiple `Ceed` objects and access them from threads later. What if we used a mutex/critical section for registration so any threads can call it, the run-time or app might choose to join those threads and create new ones, and each thread would still be able to safely use any number of `Ceed`s?

Among other things, one thread could call `CeedInit` a few times for different resources and deal them out to OpenMP threads dynamically.

Reply:
Yeah that was the idea: Each thread operates on its own `Ceed` object for its full lifetime. This is basically the idea laid out here as I understood it: https://github.com/CEED/libCEED/issues/1163#issuecomment-1438955698.

The reason for ending up with the current one thread--one `Ceed` design after kind of attempting to support the idea you state above was because it seemed to get very messy very quickly to ensure thread safety with multiple threads accessing a single `Ceed`. It seemed like `omp critical`s would start to pop up everywhere or else we would just rely on the user to not create race conditions (reference counting on a `Ceed` object, for example). So, enforcing that each thread owns its own `Ceed` seems safest to me, with a minimal amount of code changes and a fully clean `-fsanitize=thread` report. I don't think it hurts the user too badly either, since this is meant to be an alternative to MPI parallelism where you have a hard wall between `Ceed`s owned by different MPI processes. This PR just allows those `Ceed`s to all live in the same address space.

The use case I have in mind and have implemented is for the app to manage a `threadprivate` `Ceed` as:

```
Ceed ceed;
#pragma omp threadprivate(ceed)

...

void f() {
  #pragma omp parallel
  {
    CeedInit(ceed_spec, &ceed);
  }
}
```

In the example you describe, where one thread calls `CeedInit` and doles them out later on to each thread, I think that one thread will still need to call `CeedInit` once for each thread which may be spawned later on because of the issues associated with threads sharing a `Ceed` object above. In this case I don't believe the above code proposes much of a limitation.

Perhaps it's easier to discuss this over a call to work out the best solution, I am happy to adjust the implementation but as is I'm not sure it would be so easy to change and ensure thread safety.

Reply:
I can see how it feels like it's getting messy. I'm not sure it's that bad.

Per the OpenMP-5 standard, it's very delicate whether both `omp parallel` blocks would have the same thread private variables.

```c
Ceed ceed;
#pragma omp threadprivate(ceed)

#pragma omp parallel
{
  CeedInit(ceed_spec, &ceed);
}
// do some serial stuff
#pragma omp parallel
{
  CeedVectorCreate(ceed, ...);
}
```

![image](https://github.com/CEED/libCEED/assets/3303/7dd7a7f5-d92a-4545-bb76-8bddadde7292)

This includes implementation-defined behavior if the user has not exported `OMP_DYNAMIC=false`, which just feels too fragile. I think there are a couple options:

1. We could do registration into a local variable an `CeedInit` and free that list. This should eliminate global variables entirely.
2. We can have a mutex around registration (like a `once_cell`)

In either case, I think the rule is that only one thread can call methods on any given `Ceed` or object at the same time. You can "give" a `Ceed` (or other object) to a different thread, but the caller is responsible for the memory ordering necessary to ensure the last task on the donor thread completes before the recipient thread uses it.

Reply:
To make sure I understand, is the vision that the application code might configure an array of `Ceed` objects with no private data in any of them such that any one can be accessed by any worker thread? Then a parallel region could look like:

```
#pragma omp parallel
{
  int i = omp_get_thread_num();
  CeedVectorCreate(ceed[i], ...);
}
```

Reply:
Yeah, or even a higher level library (above libCEED) wants to write an interface `TakeOneStep(...)` that is meant to be called outside a parallel region. So it'll have an `omp parallel` region inside and that should be reliable and not dependent on globals.

Reply:
OK, cool that makes sense. Just in the interest of saving time, do you have a vision of how you'd make backend and Q-function registration (the two places `#pragma omp threadprivate` is currently required) into local varaibles only? These global variables (the `backends` and `gallery_qfunctions` arrays, along with `register_all_called` variables for each), are shared across various initialization functions. Where would they be kept if local?

This seems like a big change in the handling of initialization and the registration "state". Alternatively the approach of keeping the globals but adding a mutex would just involve adding a critical around use of `register_all_called` in `ceed-register.c` and `ceed-qfunction-register.c`, in addition to `critial`s or a mutex around the global `backends` and `gallery_qfunctions` variables when they are being modified. Maybe the second approach actually does sound simpler, as you were saying?